sumologic/async.py

```diff
@@ -0,0 +1,185 @@
+
+    async def collector(self, collector_id):
+        async with r:
+            r = await self.get('/collectors/' + str(collector_id))
+            return json.loads(await r.text()), r.headers['etag']
+
+    async def update_collector(self, collector, etag):
+        headers = {'If-Match': etag}
+        return await self.put('/collectors/' + str(collector['collector']['id']), collector, headers)
+
+    async def delete_collector(self, collector):
+        return await self.delete('/collectors/' + str(collector['id']))
+
+    async def sources(self, collector_id, limit=None, offset=None):
+        params = {'limit': limit, 'offset': offset}
+        async with r:
+            r = await self.get('/collectors/' + str(collector_id) + '/sources', params)
+            return json.loads(await r.text())['sources']
+
+    async def source(self, collector_id, source_id):
+        async with r:
+            r = await self.get('/collectors/' + str(collector_id) + '/sources/' + str(source_id))
+            return json.loads(await r.text()), r.headers['etag']
+
+    async def create_source(self, collector_id, source):
+        return await self.post('/collectors/' + str(collector_id) + '/sources', source)
+
+    async def update_source(self, collector_id, source, etag):
+        headers = {'If-Match': etag}
+        return await self.put('/collectors/' + str(collector_id) + '/sources/' + str(source['source']['id']), source, headers)
+
+    async def delete_source(self, collector_id, source):
+        return await self.delete('/collectors/' + str(collector_id) + '/sources/' + str(source['source']['id']))
+
+    async def create_content(self, path, data):
+        r = await self.post('/content/' + path, data)
+        async with r:
+            return await r.text()
+
+    async def get_content(self, path):
+        r = await self.get('/content/' + path)
+        async with r:
+            return json.loads(await r.text())
+
+    async def delete_content(self):
+        r = await self.delete('/content/' + path)
+        async with r:
+            return json.loads(await r.text())
+
+    async def dashboards(self, monitors=False):
+        params = {'monitors': monitors}
+        r = await self.get('/dashboards', params)
+        async with r:
+            return json.loads(await r.text())['dashboards']
+
+    async def dashboard(self, dashboard_id):
+        r = await self.get('/dashboard' + str(dashboard_id))
+        async with r:
+            return json.loads(await r.text())['dashboard']
+
+    async def dashboard_data(self, dashboard_id):
+        r = await self.get('/dashboards/' + str(dashboard_id) + '/data')
+        async with r:
+            return json.loads(await r.text())['dashboardMonitorDatas']
+
+    async def search_metrics(self, query, fromTime=None, toTime=None, requestedDataPoints=600, maxDataPoints=800):
+        '''Perform a single Sumo metrics query'''
+
+        def millisectimestamp(ts):
```

Comment:
I believe the common functions should be extracted to a util namespace to reduce code duplications.